### DISCUSSION TITLE
Generate Liquid based on AST

### DISCUSSION THREAD

**Discussion by axelv** [jg-rp/liquid#161] [Q&A] [answered]
High @jg-rp, thx for the awesome library you've made 🙌 .

Is it possible to _generate_ a Liquid template using the AST?

**Comment by jg-rp** [comment]
Hi @axelv,

Not currently.

If you're looking to build an AST programatically, it would be tricky to implement and fairly horrible to use for all but the most simple of Liquid expressions. Perhaps a builder UI would be needed to make such a thing manageable.

If you already have a `BoundTemplate`, the string representation of `BoundTemplate.tree` is not terribly useful. There's no good reason for this and I'd be happy to fix it.

```python
from liquid import Template

template = Template(
    """\
<head>
    <title>{{ site.name }} - {{ page.title }}</title>
</head>
<body>
    {% for x in (1..3) %}
      <h1>Hello, {{ you }}</h1>
    {% endfor %}
</body>"""
)

print(template.tree)
```

**output (not valid Liquid)**

```plain
<head>
    <title>`site.name` - `page.title`</title>
</head>
<body>
    for (x in (1..3)) { 
      <h1>Hello, `you`</h1>
     }
</body>
```

**Reply by axelv** [reply]
Thx for the quick response! 

Indeed I'm exploring the possibility to modify a Liquid-template programmatically and then serialize it back for human inspection.
We still need to do some research on our side. So this is only exploration. 

But do I understand correctly this is technically possible and with some help from our side we could work on this?

**Reply by jg-rp** [reply, marked as answer]
> But do I understand correctly this is technically possible and with some help from our side we could work on this?

Correct. It's possible, just fiddly.

Or maybe less fiddly depending on the modifications you need to make. I can imagine custom tag implementations (including custom versions of standard tags) that manipulate AST nodes at parse time rather than trying to change the AST after it's built.

**Comment by jg-rp** [comment]
FYI. [Python Liquid2](https://github.com/jg-rp/python-liquid2) is now released. It already supports template serialization with `str(template)`, plus other [new features](https://jg-rp.github.io/python-liquid2/migration/#new-features).
